src/main/java/com/ldtteam/structurize/client/gui/WindowScan.java

```diff
@@ -559,23 +567,38 @@ public int getElementCount()
             @Override
             public void updateElement(final int index, final Pane rowPane)
             {
-                final EntityType entity = tempEntities.get(index);
-                ItemStack entityIcon = entity.create(Minecraft.getInstance().level).getPickResult();
-                if (entity == EntityType.GLOW_ITEM_FRAME)
-                {
-                    entityIcon = new ItemStack(Items.GLOW_ITEM_FRAME);
-                }
-                else if (entity == EntityType.ITEM_FRAME)
+                final List<Entity> entities = tempEntities.get(index);
+                boolean handled = false;
+                EntityType<?> entityType = entities.get(0).getType();
+                Entity tempEntity = entityType.create(entities.get(0).level());
+                ItemStack entityIcon = tempEntity.getPickResult();
+                List<ItemStorage> requiredItems = List.of();
+                for (final IEntityHandler entityHandler : EntityHandlers.handlers)
                 {
-                    entityIcon = new ItemStack(Items.ITEM_FRAME);
+                    if (entityHandler.canHandle(entities.get(0)))
+                    {
+                        handled = entityHandler.canPlace(entities.get(0), false, true);
+                        entityIcon = entityHandler.getIcon(tempEntity);
+
+                        // unique items in descending order of required count
+                        requiredItems = entities.stream()
+                                .flatMap(e -> entityHandler.getRequiredItems(e).stream())
+                                .filter(s -> !s.isEmpty()).map(ItemStorage::new)
+                                .collect(Collectors.toMap(s -> s, s -> s, (s1, s2) -> new ItemStorage(s1.getItemStack(), s1.getAmount() + s2.getAmount(), false)))
+                                .values().stream().sorted(Comparator.comparingInt(ItemStorage::getAmount).reversed())
+                                .toList();
+                        break;
+                    }
                 }
-                else if (entity == EntityType.MINECART)
+                rowPane.findPaneOfTypeByID(RESOURCE_QUANTITY_MISSING, Text.class).setText(Component.literal(Integer.toString(entities.size())));
+                rowPane.findPaneOfTypeByID(RESOURCE_ICON, ItemIcon.class).setItem(entityIcon);
+                rowPane.findPaneOfTypeByID(RESOURCE_NAME, Text.class).setText(entityType.getDescription());
+                for (int i = 1; i <= 4; ++i)
```

Comment:
nitpick - me basically don't care, it's just more common to use this when accessing arrays/lists
```suggestion
                for (int i = 0; i < 4; ++i)
```

Reply:
That's fair.  It seemed simpler to do it this way to keep the ids 1-based, but I don't mind changing this.